### ISSUE TITLE
Simplifying setup environment variables

### ISSUE THREAD

**Issue by seancolsen** [mathesar-foundation/mathesar#3355]
### Discussed in https://github.com/mathesar-foundation/mathesar/discussions/3347

<div type='discussions-op-text'>

<sup>Originally posted by **domdomegg** December 11, 2023</sup>
I think it should be easy as possible to get Mathesar up and running in an appropriate configuration. This would help gain new users.

One of the steps on the critical path to onboarding is having to set envrionment variables. Currently there are four required environment variables: https://docs.mathesar.org/configuration/env-variables/. Each of these is an opportunity to make a mistake, and adds complexity to the setup process.

I think we can narrow it down to just the DJANGO_DATABASE_URL:

- ALLOWED_HOSTS to default to `*`, so it can be removed from required configuration. This may require a review of the code to check where we might return the host header or make decisions on it, to ensure this does not leave the default configuration to [Host header attacks](https://docs.djangoproject.com/en/4.2/topics/security/#host-headers-virtual-hosting) - however, it's reasonably likely that this is possible to do.
- MATHESAR_DATABASES to be removed and replaced by the UI connection manager (i.e. see #3223, #3344)
- SECRET_KEY to be automatically generated on startup, and then stored in the database</div>

**Comment by seancolsen** [comment]
@mathemancer and @Anish9901 I've assigned this issue to you because I'd like one or both of you to comment with your thoughts on @domdomegg's proposal.

He wants us to:

- Make `ALLOWED_HOSTS` optional
- Auto-generate `SECRET_KEY`

Can we do that?

**Comment by Anish9901** [comment]
Simplifying setting up environment variables is exactly what we are trying to achieve with our docker compose setup in the upcoming release!

https://github.com/mathesar-foundation/mathesar/blob/e9b3226a1edf1f4baebc1c975420e0f5193e93bb/docker-compose.yml#L16-L29
As you can read from the comments next to the variables only the `SECRET_KEY` is required to be added by the user the comment also suggests a website to generate one, other env variables are completely optional for the setup, however can be customised to suit your needs.

Regarding autogenerating secret key:
- I don't think storing `SECRET_KEY` on a database would be a great idea for the security reasons [django](https://docs.djangoproject.com/en/3.2/ref/settings/#secret-key) enlists.
- We could generate it on fly every time the server restarts but that would invalidate all the sessions and cookies and users would have to log in again everytime the server restarts.

**Comment by domdomegg** [comment]
I agree regenerating the `SECRET_KEY` on each restart in a production environment seems like a bad idea given this would invalidate the sessions and require users to login in again.

However, I think it still might be worth further exploring:
1. Would this be an acceptable setup for a non-production environment, e.g. to get a service up and running asap. But maybe this would create more confusion than it's worth to add this complexity.
2. What the exact reasons for not storing it in the database are. I had a read over the link you provided and didn't find anything about storage of this secret. But agree in principle that storing it in a database without knowing the security protections around the database seems potentially risky - maybe this is sufficient to rule out this idea for now.

**Comment by mathemancer** [comment]
@domdomegg for the "get it up and running ASAP" use-case, you will be able to use the single docker image:
```
docker run \
  --detach \
  -v static:/code/static \
  -v media:/code/media \
  -v postgresql_config:/etc/postgresql/ \
  -v postgresql_data:/var/lib/postgresql/ \
  --name mathesar_service \
  -p 8000:8000 \
  --restart unless-stopped \
  mathesar/mathesar-prod:latest
```
This will work as given with the next release 0.1.4 (for now, you'd need to build a local image from `develop` and tag it). I don't see a use-case for a non-prod setup that is different than that one, but which accepts mishandling the `SECRET_KEY`.

Also, we do have a script in the project that can be used to generate a secret key for production use. It absolutely doesn't work for this project to generate a new secret key on each restart; it's used to encrypt some fields on the Django DB.

In the end:  Auto-generating it in production would need to be one-shot, and it'd need to be stored in the docker compose (or environment) file afterwards. I don't think it's a big ask to have the user use the provided script. For non-production local use, it's a non-issue since you can just use our default hard-coded key.

Given that the allowed hosts issue has already been resolved as well, I think we should close this issue.

**Comment by kgodey** [comment]
>  I don't think it's a big ask to have the user use the provided script.

@mathemancer Can we make the script more discoverable in the documentation?

**Comment by seancolsen** [comment]
I want to piggy-back on @kgodey's comment...

Regarding `SECRET_KEY`, it seems likely that we have the following constraints:

- We can't store it in the DB, because there might be security risks.
- We can't auto-generate it because Mathesar doesn't necessarily have access to modify its environment variables.
- We supply a hard-coded `SECRET_KEY` value within our unified Docker image, but for security it's important that administrators deploying _production_ Mathesar take steps to re-generate this value.

I can see how these constraints make it hard to consider @domdomegg's proposals, but I think it's still worth thinking creatively about other ways that we might be able to alleviate the underlying problem.

For example, what is the failure behavior of Mathesar when an administrator does not supply a `SECRET_KEY` environment variable? (I would try this myself to answer the question, but I'm not sure how to do this without having the Docker images pushed — sorry.) If possible, I think it would be helpful for this thread if @mathemancer or @Anish9901 could post a screenshot that demonstrates this failure behavior we expect for Mathesar 0.1.4. This way we can assess the extent of the friction that administrators will encounter.

I wonder if we might have some opportunity to reduce friction by providing a more graceful failure experience. The ideal failure behavior I'm imagining is that Mathesar refuses to function normally and prints a message like this, either in the terminal or in the UI:

> **Error: Missing `SECRET_KEY` environment variable**
>
> Mathesar needs the [`SECRET_KEY` environment variable](https://docs.mathesar.org/configuration/env-variables/#secret_key) to be configured before it can start.
>
> - If you are supplying environment variables **through a `.env` file**, then run this command:
>
>     ```
>     ./setup/generate_secret_key >> .env
>     ```
>
>     That will generate a new `SECRET_KEY` value and save it to your `.env` file.
>
> - If you are supplying environment variables directly **to a `docker run` command**, then:
>
>     1. Run `./setup/generate_secret_key` and copy the output.
>     1. Paste the output after the `--env` flag.
>     1. Be sure to supply the _same_ `SECRET_KEY` value any time you restart Mathesar, otherwise your PostgreSQL connection credentials and user sessions will be invalidated.


I'd like to point out that in my suggestion above relies on an executable script `./setup/generate_secret_key` which outputs something like this:

```
SECRET_KEY=jm4OeNmIPwrCjI4rIWQS2ehTaDPeDLReb5nKiR0GemY
```

That's different than the current script I see at `setup/secret_key_generator.py` which outputs something like this:

```
Please follow the instructions in <Add documentation link> to add the below secret key to the application
Secret Key: jm4OeNmIPwrCjI4rIWQS2ehTaDPeDLReb5nKiR0GemY
```

Basically at a high level, what I'm after here is an assessment of the finer points of the administrator's experience that we expect in 0.1.4 with regard to `SECRET_KEY`

Overall, I'm trying to find ways to reduce friction, even within the confines of our constraints.

**Comment by seancolsen** [comment]
I've re-labeled this issue as `needs: requirements` since it's a more open-ended problem now, instead of a specific proposal.

**Comment by mathemancer** [comment]
@seancolsen The intended behavior for 0.1.4 is that the admin runs `sudo docker compose -f docker-compose.yml up` as instructed within the file itself. If they haven't replaced the `SECRET_KEY` variable in that file, it prints:

```
parsing /home/brent/dev/cci/repos/mathesar/docker-compose.yml: error while interpolating x-config.SECRET_KEY: required variable SECRET_KEY is missing a value
```
If they then look at that variable in the `docker-compose.yml` file (which they should have done before running it anyway), they see:

```
  # (Required) Replace '?' with '-' followed by your 50 character long Django
  # SECRET_KEY, you can generate one here: https://djecrety.ir/
```

This issue has been resolved at least as far as makes any sense, and should be closed.

**Comment by mathemancer** [comment]
> > I don't think it's a big ask to have the user use the provided script.
> 
> @mathemancer Can we make the script more discoverable in the documentation?

I was wrong to mention the script.  That only comes with the whole repo, and doesn't really make sense in the context in which this file is intended to be used. See my comment above.

**Comment by seancolsen** [comment]
Brief update on this issue:

We discussed this during a maintainers meeting today.

**Next steps:** We're going to wait until the docs are further along and then evaluate the experience of administrators from start to finish regards to SECRET_KEY to see how smooth the process is. We'll follow up on this thread at that point, hoping to demonstrate a smooth experience and get more feedback as necessary.

**Comment by seancolsen** [comment]
Update! [Mathesar 0.1.4](https://docs.mathesar.org/releases/0.1.4/) is now released.

Mathesar _does_ need the `SECRET_KEY` variable to be configured during installation. But we've made improvements to to the installation process and documentation in order to reduce friction here.

- When [installing Mathesar via Docker Compose](https://docs.mathesar.org/installation/docker-compose/), our docs give a one-line command to generate a `SECRET_KEY` value that can be pasted into docker-compose.yml

    ![image](https://github.com/mathesar-foundation/mathesar/assets/42411/825b0e4c-64ea-485c-a714-d983bc0ac2e8)

- When [installing Mathesar from scratch](https://docs.mathesar.org/installation/build-from-source/), we have a similar step:

    ![image](https://github.com/mathesar-foundation/mathesar/assets/42411/b79a4c31-5546-4a7a-b2a0-b800b4a464f2)

@domdomegg if you have any other feedback, or suggestions for how we could improve this, we'd love to hear it!

If we don't see any further comments on this thread by 2024-02-14 we'll consider this resolved and close this issue.
